crypto/mem_sec.c

```diff
@@ -433,8 +438,19 @@ static int sh_init(size_t size, int minsize)
     if (mprotect(sh.map_result + aligned, pgsize, PROT_NONE) < 0)
         ret = 2;
 
+#if defined(OPENSSL_SYS_LINUX) && defined(MLOCK_ONFAULT) && defined(SYS_mlock2)
+    if (syscall(SYS_mlock2, sh.arena, sh.arena_size, MLOCK_ONFAULT) < 0) {
+        if (errno == ENOSYS) {
+            if (mlock(sh.arena, sh.arena_size) < 0)
+                ret = 2;
+        } else {
```

Comment:
This by the way calls for following question. Given description of the flag used with mlock2, mlock might do nothing under the circumstances. I mean we just did anonymous mapping and haven't used it in any way, so they might be actually *no* "wired" pages yet. And if mlock indeed affects only "wired" pages, then it effectively does nothing. Maybe it would be appropriate to touch each page prior mlock? As for failures. There is resource limit for how much memory you can lock. [And ability to increase it is a question for privilege.] With this in mind is it possible to that mlock can succeed where mlock2 would fail? But if it's possibility, then it would be proof that we really need to touch each page prior mlock...